File: fastapi/app/checkout.py
import os
import re
from dataclasses import dataclass
from urllib.parse import urlparse

import inttegro
from inttegro import APIError, LineItemType, ProductType


# Inttegro integration map
#
# INTTEGRO:FLOW [hosted-checkout] This server-only module validates an untrusted
# ticket request, creates and finalizes an Order, and returns the hosted invoice
# URL used by the FastAPI route.
# INTTEGRO:SECURITY [server-api-key] INTTEGRO_API_KEY stays in the trusted Python
# process. Never publish it through frontend configuration, HTML, client code,
# mobile bundles, or error responses.
# INTTEGRO:ALTERNATIVE [hosted-checkout] Direct API payment offers additional UI
# control but makes the merchant own more payment states, recovery behavior,
# method-specific logic, testing, and compliance analysis.
# INTTEGRO:DOCS https://studio.inttegro.com/accept-payment-with-inttegro-checkout
# INTTEGRO:DOCS https://studio.inttegro.com/orders
# INTTEGRO:DOCS https://studio.inttegro.com/keys
# See ../../INTEGRATION_GUIDE.md and ../../integration-decisions.json for the full
# rationale and machine-readable alternatives.


class DemoError(Exception):
    def __init__(self, code: str, message: str) -> None:
        super().__init__(message)
        self.code = code
        self.message = message


@dataclass(frozen=True)
class CheckoutInput:
    name: str
    email: str
    phone: str
    attempt_id: str


@dataclass(frozen=True)
class CatalogSelection:
    type: ProductType
    name: str
    about: str | None
    reference: str | None
    price: inttegro.PriceParams


def parse_checkout_input(name: str, email: str, phone: str, attempt_id: str) -> CheckoutInput:
    value = CheckoutInput(name.strip(), email.strip(), phone.strip(), attempt_id.strip())
    if not value.name or not value.phone or not re.fullmatch(r"[^\s@]+@[^\s@]+\.[^\s@]+", value.email):
        raise DemoError("validation_error", "Enter a name, valid email, and phone number.")
    if not re.fullmatch(r"[A-Za-z0-9_-]{8,100}", value.attempt_id):
        raise DemoError("validation_error", "Start a fresh checkout and try again.")
    return value


def validated_origin(value: str) -> str:
    # INTTEGRO:SECURITY [configured-public-origin] Production should use an
    # allow-listed configured origin. A request-derived fallback is safe only
    # behind a precisely configured trusted proxy chain. Return URLs must be
    # absolute HTTP(S); Inttegro appends order_id after checkout.
    parsed = urlparse(value)
    if parsed.scheme not in {"http", "https"} or not parsed.netloc:
        raise DemoError("configuration_error", "The demo public URL is invalid.")
    return f"{parsed.scheme}://{parsed.netloc}"


def select_catalog_product(product, price_id: str) -> CatalogSelection:
    # INTTEGRO:SECURITY [catalog-authority] The untrusted form supplies no
    # product, price, or amount. Resolve and validate them on this server.
    if not product.active:
        raise DemoError("configuration_error", "The configured demo product is not active.")
    price = next((candidate for candidate in (product.prices or []) if candidate.id == price_id and candidate.active), None)
    if price is None or price.nominal.value <= 0:
        raise DemoError("configuration_error", "The configured demo price is not active for this product.")
    return CatalogSelection(
        type=product.type,
        name=product.name,
        about=product.about,
        reference=product.reference,
        price=inttegro.PriceParams(currency=price.nominal.currency, value=price.nominal.value),
    )


def build_order_request(value: CheckoutInput, origin: str, product: CatalogSelection) -> inttegro.orders.CreateRequest:
    return inttegro.orders.CreateRequest(
        # INTTEGRO:DECISION [stable-idempotency-key] Reuse the key for retries of
        # this logical checkout. Production systems persist a cart- or booking-
        # derived key; creating a fresh key after a timeout permits duplicates.
        # https://studio.inttegro.com/idempotency
        request_meta=inttegro.orders.RequestMeta(idempotency_key=f"demo-{value.attempt_id}"),
        # INTTEGRO:DECISION [inline-customer] customer_data matches this guest
        # flow. Account-based products should resolve customer_id on the trusted
        # server. The Orders API accepts exactly one customer representation.
        customer_data=inttegro.orders.Customer(name=value.name, email_address=value.email, phone_number=value.phone),
        # INTTEGRO:DECISION [finalize-on-create] The item and total are settled,
        # so finalization freezes the order and generates invoice formats in one
        # call. Use draft -> update -> finalize for mutable tax, inventory,
        # approval, shipping, or line items.
        finalize=True,
        checkout_settings=inttegro.orders.CheckoutSettings(redirect_url=f"{origin}/complete", cancel_url=f"{origin}/cancel"),
        line_items=[inttegro.orders.ProductLineItem(
            type=LineItemType.PRODUCT,
            product=inttegro.orders.Product(
                # INTTEGRO:DECISION [catalog-snapshot] The service resolves the
                # configured Product and Price and snapshots the verified fields
                # into the Order for compatibility across SDK versions.
                # INTTEGRO:ALTERNATIVE [catalog-snapshot] SDKs with the typed
                # catalogue union can send product_id + price_id + quantity
                # instead, without mixing in inline product fields.
                type=product.type,
                name=product.name,
                about=product.about,
                reference=product.reference,
                quantity=1,
                price=product.price,
            ),
        )],
    )


def create_hosted_checkout(value: CheckoutInput, origin: str) -> tuple[str, str]:
    # INTTEGRO:SECURITY [server-api-key] Source this value from a managed runtime
    # secret in production; never include it in logs or client-visible errors.
    api_key = os.environ.get("INTTEGRO_API_KEY", "").strip()
    if not api_key:
        raise DemoError("configuration_error", "Set INTTEGRO_API_KEY on the server.")
    product_id = os.environ.get("INTTEGRO_DEMO_PRODUCT_ID", "").strip()
    price_id = os.environ.get("INTTEGRO_DEMO_PRICE_ID", "").strip()
    if not re.fullmatch(r"prod_[A-Za-z0-9]+", product_id) or not re.fullmatch(r"pr_[A-Za-z0-9]+", price_id):
        raise DemoError("configuration_error", "Set INTTEGRO_DEMO_PRODUCT_ID and INTTEGRO_DEMO_PRICE_ID on the server.")
    try:
        # INTTEGRO:ALTERNATIVE [server-api-key] Long-running apps can inject one
        # startup-configured client for connection reuse and application-owned
        # OpenTelemetry. The SDK chooses no exporter or vendor.
        # https://studio.inttegro.com/sdk-observability
        client = inttegro.InttegroClient(api_key=api_key)
        # INTTEGRO:FLOW [catalog-lookup] Fetch at checkout time so publication
        # and price changes are observed. High-volume apps may use a short cache
        # with an explicit invalidation policy. https://studio.inttegro.com/products
        product = select_catalog_product(client.products.lookup(product_id), price_id)
        order = client.orders.create(build_order_request(value, validated_origin(origin), product))
        # INTTEGRO:DECISION [returned-checkout-url] Use the response field rather
        # than constructing a URL from order.id and undocumented routing.
        checkout_url = order.invoice.format.web.url
        if not checkout_url:
            raise DemoError("api_error", "Inttegro did not return a hosted checkout URL.")
        return order.id, checkout_url
    except DemoError:
        raise
    except APIError as error:
        # INTTEGRO:SECURITY [safe-error-boundary] Keep raw upstream payloads,
        # credentials, and traces private. Log bounded metadata and request IDs;
        # expose only stable public categories.
        raise DemoError("api_error", "Inttegro rejected the checkout request.") from error
    except Exception as error:
        raise DemoError("api_error", "Checkout is temporarily unavailable.") from error

File: nuxt/server/utils/checkout.ts
import { InttegroAPIError, InttegroClient, type CreateOrderRequest, type Product } from '@inttegro/inttegro-sdk';

/**
 * Inttegro integration map
 *
 * INTTEGRO:FLOW [hosted-checkout] This Nitro server utility validates an
 * untrusted form, creates a finalized Order, and returns the hosted invoice URL
 * that the route will redirect to.
 * INTTEGRO:SECURITY [server-api-key] Runtime config must keep the key private;
 * never place it under runtimeConfig.public or a NUXT_PUBLIC_* variable.
 * INTTEGRO:ALTERNATIVE [hosted-checkout] A direct API payment flow is possible,
 * but the merchant then owns additional payment states, recovery UI,
 * method-specific behavior, testing, and compliance analysis.
 * INTTEGRO:DOCS https://studio.inttegro.com/accept-payment-with-inttegro-checkout
 * INTTEGRO:DOCS https://studio.inttegro.com/orders
 * INTTEGRO:DOCS https://studio.inttegro.com/keys
 * See ../../../INTEGRATION_GUIDE.md and ../../../integration-decisions.json for the shared
 * rationale and machine-readable alternatives.
 */

export class DemoError extends Error {
  constructor(readonly code: 'configuration_error' | 'validation_error' | 'api_error', message: string) { super(message); }
}

export type CheckoutInput = { name: string; email: string; phone: string; attemptId: string };

export type CatalogSelection = {
  type: Product['type'];
  name: string;
  about?: string;
  reference?: string;
  price: NonNullable<Product['prices']>[number]['nominal'];
};

function merchantOrderNumber(prefix: string, attemptId: string): string {
  return `${prefix}-${attemptId.replaceAll('_', '-').toUpperCase().slice(0, 48)}`;
}

export function parseCheckoutInput(body: Record<string, unknown>): CheckoutInput {
  const input = { name: String(body.name ?? '').trim(), email: String(body.email ?? '').trim(), phone: String(body.phone ?? '').trim(), attemptId: String(body.attempt_id ?? '').trim() };
  if (!input.name || !input.phone || !/^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(input.email)) throw new DemoError('validation_error', 'Enter a name, valid email, and phone number.');
  if (!/^[A-Za-z0-9_-]{8,100}$/.test(input.attemptId)) throw new DemoError('validation_error', 'Start a fresh checkout and try again.');
  return input;
}

export function selectCatalogProduct(product: Product, priceId: string): CatalogSelection {
  // INTTEGRO:SECURITY [catalog-authority] Resolve the Product and Price on this
  // trusted Nitro server. The browser sends neither identifier nor amount, so
  // it cannot swap catalogue records or tamper with the price.
  if (!product.active) throw new DemoError('configuration_error', 'The configured demo product is not active.');
  const price = product.prices?.find(candidate => candidate.id === priceId && candidate.active);
  if (!price || !Number.isSafeInteger(price.nominal.value) || price.nominal.value <= 0) {
    throw new DemoError('configuration_error', 'The configured demo price is not active for this product.');
  }
  return {
    type: product.type,
    name: product.name,
    ...(product.about ? { about: product.about } : {}),
    ...(product.reference ? { reference: product.reference } : {}),
    price: price.nominal,
  };
}

export function buildOrderRequest(input: CheckoutInput, origin: string, product: CatalogSelection): CreateOrderRequest {
  return {
    // INTTEGRO:DECISION [stable-idempotency-key] Reuse this key when retrying the
    // same logical checkout. Production systems should persist a cart- or
    // invoice-derived key; a fresh random key per network retry permits duplicates.
    // https://studio.inttegro.com/idempotency
    request_meta: { idempotency_key: `demo-${input.attemptId}` },
    // INTTEGRO:DECISION [merchant-order-number] Supply a recognizable merchant
    // reference rather than accepting Inttegro's generated or_ ID fallback. The
    // demo derives it from the validated attempt so replayed requests agree.
    // Production systems should use a persisted cart/order number without PII.
    number: merchantOrderNumber('KORA', input.attemptId),
    // INTTEGRO:DECISION [inline-customer] customer_data fits this guest flow.
    // Account-based products should resolve customer_id on the server; the
    // Orders API accepts exactly one of these representations.
    customer_data: { name: input.name, email_address: input.email, phone_number: input.phone },
    // INTTEGRO:DECISION [finalize-on-create] This fixed cart is complete, so one
    // call freezes the order and creates invoice formats. If shipping, tax,
    // inventory, or approval is pending, create a draft and finalize later.
    finalize: true,
    checkout_settings: { redirect_url: `${origin}/complete`, cancel_url: `${origin}/cancel` },
    line_items: [{
      type: 'product',
      product: {
        // INTTEGRO:DECISION [catalog-snapshot] The service looks up the configured
        // Inttegro Product and Price, validates their relationship, and snapshots
        // that authoritative data into the Order. This works across SDK versions.
        // INTTEGRO:ALTERNATIVE [catalog-snapshot] SDKs that expose the typed
        // catalogue union can instead send product_id + price_id + quantity.
        // Never mix that reference shape with these inline product fields.
        ...product,
        quantity: 1,
      },
    }],
  };
}

export async function createHostedCheckout(
  input: CheckoutInput,
  apiKey: string,
  origin: string,
  productId: string,
  priceId: string,
) {
  // INTTEGRO:SECURITY [server-api-key] The SDK call belongs in Nitro server
  // code. Use a deployment secret manager in production, not source or public
  // runtime configuration.
  if (!apiKey.trim()) throw new DemoError('configuration_error', 'Set NUXT_INTTEGRO_API_KEY on the server.');
  if (!/^prod_[A-Za-z0-9]+$/.test(productId) || !/^pr_[A-Za-z0-9]+$/.test(priceId)) {
    throw new DemoError('configuration_error', 'Set NUXT_DEMO_PRODUCT_ID and NUXT_DEMO_PRICE_ID on the server.');
  }
  let publicOrigin: string;
  // INTTEGRO:SECURITY [configured-public-origin] Prefer an allow-listed origin
  // in production. A request-derived fallback is safe only behind a precisely
  // configured trusted proxy chain. Return URLs must be absolute HTTP(S) URLs.
  try { publicOrigin = new URL(origin).origin; } catch { throw new DemoError('configuration_error', 'The demo public URL is invalid.'); }
  try {
    // INTTEGRO:ALTERNATIVE [server-api-key] Long-running apps may inject one SDK
    // client configured at startup for connection reuse and application-owned
    // OpenTelemetry. Per-operation construction keeps this demo focused.
    // https://studio.inttegro.com/sdk-observability
    const client = new InttegroClient({ apiKey });
    // INTTEGRO:FLOW [catalog-lookup] Fetch at checkout time so publication and
    // price changes are observed. A high-volume app may use a short bounded cache
    // with explicit invalidation rather than an unbounded process-lifetime cache.
    // https://studio.inttegro.com/products
    const product = selectCatalogProduct(await client.products.lookup({ product_id: productId }), priceId);
    const order = await client.orders.create(buildOrderRequest(input, publicOrigin, product));
    // INTTEGRO:DECISION [returned-checkout-url] Use the URL in the response;
    // constructing one from order.id relies on undocumented routing details.
    const checkoutUrl = order.invoice?.format?.web?.url;
    if (!checkoutUrl) throw new DemoError('api_error', 'Inttegro did not return a hosted checkout URL.');
    return { checkoutUrl, orderId: order.id };
  } catch (error) {
    // INTTEGRO:SECURITY [safe-error-boundary] Never expose raw SDK payloads,
    // authorization data, or stack traces. Keep request IDs and safe diagnostics
    // in access-controlled server telemetry.
    if (error instanceof DemoError) throw error;
    if (error instanceof InttegroAPIError) throw new DemoError('api_error', 'Inttegro rejected the checkout request.');
    throw new DemoError('api_error', 'Checkout is temporarily unavailable.');
  }
}
